### ISSUE TITLE
Remove dependency on require.js

### ISSUE THREAD

**Issue by alxndrsn** [enketo/enketo-core#286]
Hi Martijn, not sure how feasible this is but it would be handy on client-side if `enketo-core` didn't require `requirejs`.  I'll probably take a look at this in the next few days, but would be interested to hear your thoughts on making the `requirejs` dependency optional, and if you'd accept a patch for that.

**Comment by MartijnR** [comment]
I totally agree. I'd like to use CommonJS modules instead (probably with browserify). Backwards compatibility with requirejs would be nice-to-have (and I'm sure a lot of people would be very happy), but in my opinion it is not required if this gets too complex. If it breaks AMD-compatibility, we'd just make this v4.0.0.

So yes, I'd very much welcome a PR/patch for this!

**Comment by alxndrsn** [comment]
Any thoughts on how widget config should be fetched, if not using `requirejs`?  It's the first non-obvious thing I've come to:
```javascript
//add widget configuration to config object
for ( var i = 0; i < config.widgets.length; i++ ) {
    id = 'text!' + config.widgets[ i ].substr( 0, config.widgets[ i ].lastIndexOf( '/' ) + 1 ) + 'config.json';
    widgetConfigFiles.push( id );
}

//load widget config files
require( widgetConfigFiles, function() {

    for ( var i = 0; i < arguments.length; i++ ) {
        widget = JSON.parse( arguments[ i ] );
        widget.path = config.widgets[ i ];
        widgets.push( widget );
    }

    deferred.resolve( widgets );
} );
```
> https://github.com/enketo/enketo-core/blob/master/src/js/widgets.js#L104

Are users expected to change widget configs?

And is there anything that's needed pre-loadtime apart from `options.selector`?  If not, would it work to convert enketo's `config.json` `widgets` object from `["widget-1", ... , "widget-N"]` to be a map `{"widget-1":".1-selector", ..., "widget-N": "N-selector"}`?

E.g for https://github.com/enketo/enketo-core/blob/master/config.json#L3
```json
{
    "widgets": [
        "enketo-widget/note/notewidget",
        ...
```
becomes 
```json
{
    "widgets": {
        "enketo-widget/note/notewidget": ".note",
        ...
```
I think this would remove the need for accessing individual widgets' `config.json` files from the JS.

**Comment by MartijnR** [comment]
I think your idea is good, but I'd like to keep the selector information out of the app's configuration (because it's static). I think the stylesheet info can indeed be removed if [this](https://github.com/enketo/enketo-core/blob/master/Gruntfile.js#L131-L159) becomes a more intelligent task and only includes a scss reference if the file exists.

just some background: 

The idea behind this code was to easily disable, and add widgets in a central app configuration file. Each individual widget's configuration file is expected to remain unchanged, but it points to the resources it needs (both css and js) to reduce the built file sizes by including only the widgets that are actually enabled.

I wanted to keep each widget including information on its selector and resources in its own folder, to allow people to create proprietary widgets and add these as a library without touching enketo-core code. 

If we need a central (static) widget map in addition to these folders, that's not a big deal. 

I think the ideal app config.json file would be as simple as possible, but still allow custom (non-enketo-core widgets to be added) e.g. like this:

```json
"widgets" : [
   "enketo-widget/notewidget",
   "enketo-widget/horizontal-choices",
  ...
]
```
where `enketo-widgets/` just indicates that these are enketo-widget widgets (and not custom ones). 

Although, writing this makes me wonder if we should just remove the ability to easily add custom widgets. Would make things a lot cleaner `"widgets" : ["notewidget", "horizontal-choices"]...

**Comment by MartijnR** [comment]
So how about this approach? 

CommonJS for everything: tests, app.js, modules, build system.

RequireJS (backwards) compatibility *only* for modules in /src/js. Not for tests, not for js build system. 

Next steps after CommonJS-ifying would be to make the XPath evaluator CommonJS-friendly, get rid of all git submodules and switch all bower modules to npm. (For me these are related as this will allow publishing enketo-core on npm).

How does that sound Alex?

**Comment by alxndrsn** [comment]
Sounds very reasonable.

I'm not sure about bower vs. npm, as I understood bower has additional support for css and other non-JS resources.

If the CommonJS approach is good, would it be reasonable to merge that before starting to sort out `XPath`, git submodules and bower?

**Comment by MartijnR** [comment]
Yes,  building the css files from npm modules is the one thing I'm also not so enthusiastic about.

Yes, merging commonJS before xpath, git submodules and bower would be fine. We'd just need to clean up the old requirejs stuff, fix the tests. 

Transforming the tests to use commonJS can also be done later I think (as long as they run again).

**Comment by alxndrsn** [comment]
Thanks!  I've updated our fork, and as of https://github.com/alxndrsn/enketo-core/commit/ca1bfb0012b2a6de3122cf0589d8388ef2a60d1d have separated our changes from the core enketo codebase.  Our extensions, config and widget choices are now all configured via a [separate grunt-browserify task](https://github.com/alxndrsn/enketo-core/blob/clientside-4.x/Gruntfile.js#L118).

There are a couple of changes on [this branch](https://github.com/alxndrsn/enketo-core/tree/clientside-4.x) which may be welcome on `master`:

* move module mappings from `.browser` in `package.json` to `.browserify.standalone.options.alias` in `Gruntfile.js`
* separate [`FormModel.prototype.bindJsEvaluator`](https://github.com/enketo/enketo-core/blob/master/src/js/FormModel.js#L283) into a separate module

Let me know if either is of interest, and I'll prepare a PR.

N.B. I hope this can serve as an example of _Usage as a LIbrary_ section *4* from [the readme](https://github.com/enketo/enketo-core/blob/master/README.md#L123):
> 4.Override `config.json` and optionally `widgets.js` with your app-specific versions.

If not, any pointers on what we're doing wrong would be very welcome.

**Comment by MartijnR** [comment]
Re grunt task. Yes, that's better and provides an easy to understand example of overriding config and widgets (and anything else). PR welcome! I couldn't actually get alias to work for some reason.

_FYI, I had to go a different route with enketo-express: https://github.com/kobotoolbox/enketo-express/blob/master/Gruntfile.js#L147-L155. It was the only way I could find to overwrite paths recursively (when enketo-core is added as an npm module)._

Re. separating bindJsEvaluator. PR would be useful as it provides an easy way to swap in the future. For your evaluator do you still require the native vs. custom evaluator logic (all for performance) in FormModel.prototype.evaluate? I'd be very interested in replacing the XPath evaluator with yours .... (if it passes the same tests). Do you think that is feasible from IE Edge onwards or are you initially supporting fewer XPath features perhaps?

**Comment by alxndrsn** [comment]
I found configuring browserify pretty horrible too - felt like I just kept tinkering until it randomly worked.

Ref. XPath: no idea how my evaluator matches up to others, although it definitely relies on the browser's implementation for doing standard xpath lookups.  I can have a go at getting it working against the tests; is it these ones: https://github.com/enketo/enketo-xpathjs/tree/master/test/spec?  And why do you expect it'll be faster?

**Comment by MartijnR** [comment]
Yes, those are the tests. You might run into namespace issues but I hope not.

I think yours is faster because it ends up sending far more expression fragments to the native evaluator (which is roughly 100x faster). If I understand correctly, you are parsing the expression and then decide which fragments to evaluate with javascript and which to send to native. E.g. in `count-selected(/path/to/node)`, you'd evaluate the /path/to/node argument natively. In enketo-xpathjs, the whole thing is sent to the the javascript evaluator.

**Comment by alxndrsn** [comment]
Have had a play with the tests this morning.

Definitely not going to work immediately - my XPath evaluator currently makes a big assumption - it only allows for an XPathResult.STRING_TYPE.  This is no problem for our current use, and *perhaps* not for more general Enketo use, but it certainly won't pass the tests as they stand.  E.g. nodeset-based tests like https://github.com/enketo/enketo-xpathjs/blob/master/test/spec/nodeset-functions-native.spec.js and https://github.com/enketo/enketo-xpathjs/blob/master/test/spec/union-operator-native.spec.js

Other problems I noticed, which are more easily fixed:
* parsing `and` and `or` expressions with missing spaces isn't currently supported (e.g. `1and1`)
* non-boolean `and` and `or` expressions do not follow the same spec, e.g.:
  - enketo-xpathjs: `1 and 1` -> `true`
  - openrosa-xpath-evaluator: `1 and 1` -> `1`

**Comment by MartijnR** [comment]
Interesting. Thanks for sharing this. I think union operators are not supported in ODK Validate either (so basically not used by anyone). Just note that Enketo Core does use these internally (though I think these will bypass your evaluator).

The tests cover many nooks and crannies that are never used within the ODK ecosystem (they are to a much larger extent used in CommCare though which has more robust XPath support and usage than ODK). Nodeset functions would be a problem for switching over.
